### DISCUSSION TITLE
INDIRECT function

### DISCUSSION THREAD

**Discussion by rackuka** [handsontable/hyperformula#870] [Feature requests]
Hi hyperformula team!

I've just started using hyperformula and appears the INDERECT function is not yet supported even though listed among functions to be added (https://github.com/handsontable/hyperformula/issues/221).

Do you have an idea of when the INDIRECT function can appear?

Another question is if I decide to contribute and add function myself (through PR of course), then what edge cases I should consider and pay attention to?

Presence of the INDIRECT function is really on critical path for me so I am quite motivated to have it in hyperformula.

Alexei.

**Comment by bardek8** [comment]
Hi Alexei,
in order to answer you in detail, let me briefly describe how HyperFormula works right now:
a) parse all formulas and understand from which cells a particular cell depends on
b) build a graph of dependencies between cells
c) evaluate cells in the appropriate order

Now the issue with `INDIRECT` function is that during the parsing phase (a) we cannot say on which cells the formula depends on. Then we cannot build the dependency graph properly. We would have to first evaluate the cells (c) and then we would be able to build the graph. But this requires a more dynamic approach to building the graph and evaluating the cells which means rewriting a big part of HF. Hence unfortunately we cannot provide any estimate for it.

On the positive side, we tried to handle the `OFFSET` function which has similar issues. We covered some special cases, when the arguments are static numbers and we can learn dependencies during the parsing phase. Please see details [here](https://github.com/handsontable/hyperformula/blob/b989cfde24024309eae32a3edb069de84539ddeb/src/parser/FormulaParser.ts#L781). If in your application you can detect dependencies without full evaluation, you can try a similar approach.

**Comment by rackuka** [comment]
Hi Bartek,

What you described looks like fundamental change.
Thank you for detailed clarification.

Will definitely follow approach you referred for the OFFSET heuristics.

Alexei.

**Comment by sequba** [comment]
Related: https://github.com/handsontable/hyperformula/issues/1260
